lib/ansible/modules/windows/win_domain_user.ps1

```diff
@@ -4,6 +4,109 @@
+    }
+    $null = $Domain # Make CI-Check happy...
+
+    # More information about LogonUser at https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-logonusera
+    $platform_util = @'
+using System;
+using System.Runtime.InteropServices;
+using System.ComponentModel;
+using System.Security;
+namespace Ansible
+{
+    public class WinUserPInvoke
+    {
+        const int LOGON32_LOGON_NETWORK           = 3;
+        const int LOGON32_PROVIDER_DEFAULT        = 0; // Negotiate if domain == null and username not upn else NTLM
+        const int LOGON32_PROVIDER_WINNT40        = 2; // Negotiate
+        const int LOGON32_PROVIDER_WINNT50        = 3; // NTLM
+        // Refere to: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d
+        static int[] success_codes = new int[] {
+            0x0000052F,  // ERROR_ACCOUNT_RESTRICTION
+            0x00000530,  // ERROR_INVALID_LOGON_HOURS
+            0x00000531,  // ERROR_INVALID_WORKSTATION
+            0x00000569   // ERROR_LOGON_TYPE_GRANTED
+        };
+        static int[] failed_codes = new int[] {
+            0x0000052E,  // ERROR_LOGON_FAILURE - the user or pass was incorrect
+            0x00000006,  // ERROR_INVALID_HANDLE - the handle is invalid
+            0x00000532   // ERROR_PASSWORD_EXPIRED - The password has exired
+        };
+        [DllImport("advapi32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        internal static extern bool LogonUser(String username, String domain, IntPtr password,
+            int logonType, int logonProvider, out IntPtr token);
+        [DllImport("kernel32.dll", SetLastError = true)]
+        internal static extern bool CloseHandle(
+            IntPtr hObject);
+        public static bool TestCredential(string userName, SecureString password, string domainName = null)
+        {
+            IntPtr tokenHandle = IntPtr.Zero;
+            IntPtr passwordPtr = IntPtr.Zero;
+            bool returnValue   = false;
+            int err_code = 0;
+
+            // Marshal the SecureString to unmanaged memory.
+            passwordPtr = Marshal.SecureStringToGlobalAllocUnicode(password);
+
+            // Pass LogonUser the decrypted copy of the password.
+            returnValue = LogonUser(userName, domainName, passwordPtr,
+                                    LOGON32_LOGON_NETWORK, LOGON32_PROVIDER_DEFAULT,
+                                    out tokenHandle);
+            // Perform cleanup whether or not the call succeeded.
+            // Secure delete the plaintext password from memory.
+            Marshal.ZeroFreeGlobalAllocUnicode(passwordPtr);
+            // We don't need the token, so close it again.
+            CloseHandle(tokenHandle);
+            if (returnValue || !(tokenHandle == IntPtr.Zero)) {
+                return true;
+            } else {
+                err_code = Marshal.GetLastWin32Error();
```

Comment:
You should call this straight after `LogonUser`. If you don't you run the risk of any of the preceding calls overwriting the value and then you getting the wrong error code. Right now this should always be returning `0` as the `CloseHandle` will be succeeding.